pkg/build/controller/controller.go

```diff
@@ -224,3 +224,71 @@ func (bc *BuildPodController) CancelBuild(build *buildapi.Build, pod *kapi.Pod)
 func isBuildCancellable(build *buildapi.Build) bool {
 	return build.Status == buildapi.BuildStatusNew || build.Status == buildapi.BuildStatusPending || build.Status == buildapi.BuildStatusRunning
 }
+
+// BuildPodDeleteController watches pods running builds and updates the build if the pod is deleted
+type BuildPodDeleteController struct {
+	BuildStore   cache.Store
+	BuildUpdater buildclient.BuildUpdater
+}
+
+func (bc *BuildPodDeleteController) HandleBuildPodDeletion(pod *kapi.Pod) error {
+	glog.V(4).Infof("Handling deletion of build pod %s/%s", pod.Namespace, pod.Name)
+	// pod and build have the same name, so look up the build by the pod's name.
+	obj, exists, err := bc.BuildStore.Get(pod)
+	if err != nil {
+		glog.V(4).Infof("Error getting build for pod %s/%s", pod.Namespace, pod.Name)
+		return err
+	}
+	if !exists || obj == nil {
+		glog.V(5).Infof("No build found for deleted pod %s/%s", pod.Namespace, pod.Name)
+		return nil
+	}
+	build := obj.(*buildapi.Build)
+
+	if build.Status != buildapi.BuildStatusRunning && build.Status != buildapi.BuildStatusPending && build.Status != buildapi.BuildStatusNew {
+		glog.V(4).Infof("Pod was deleted but Build %s/%s is already completed, so no need to update it.", build.Namespace, build.Name)
+		return nil
+	}
+
+	nextStatus := buildapi.BuildStatusError
+	if build.Status != nextStatus {
+		glog.V(4).Infof("Updating build %s/%s status %s -> %s", build.Namespace, build.Name, build.Status, nextStatus)
+		build.Status = nextStatus
+		build.Message = "The Pod for this Build was deleted before the Build completed."
+		dummy := util.Now()
```

Comment:
Don't use dummy, I already fixed this to use `now` in all other places in build controller. Otherwise you'll have to tackle @smarterclayton's angry eyes :wink: